pageserver/src/relish_storage/storage_sync.rs

```diff
@@ -0,0 +1,1561 @@
+    )
+    .await;
+
+    let entry_to_update = remote_timelines
+        .entry((tenant_id, timeline_id))
+        .or_insert_with(|| RemoteTimeline {
+            layers: Vec::new(),
+            metadata: Some(new_metadata.clone()),
+            tenant_id,
+            timeline_id,
+        });
+    match sync_result {
+        SyncResult::Success { synced } => {
+            entry_to_update.layers.extend(synced.into_iter());
+            entry_to_update.metadata = Some(new_metadata);
+            Some(true)
+        }
+        SyncResult::MetadataSyncError { synced } => {
+            entry_to_update.layers.extend(synced.into_iter());
+            add_to_queue(SyncTask::Upload(LocalTimeline {
+                tenant_id,
+                timeline_id,
+                layers: Vec::new(),
+                metadata: new_metadata,
+            }));
+            Some(false)
+        }
+        SyncResult::LayerSyncError { synced, not_synced } => {
+            entry_to_update.layers.extend(synced.into_iter());
+            add_to_queue(SyncTask::Upload(LocalTimeline {
+                tenant_id,
+                timeline_id,
+                layers: not_synced,
+                metadata: new_metadata,
+            }));
+            Some(false)
+        }
+    }
+}
+
+/// Layer sync operation kind.
+///
+/// This enum allows to unify the logic for image relish uploads and downloads.
+/// When image's layers are synchronized, the only difference
+/// between downloads and uploads is the [`RelishStorage`] method we need to call.
+#[derive(Debug, Copy, Clone)]
+enum SyncOperation {
+    Download,
+    Upload,
+}
+
+/// Image sync result.
+#[derive(Debug)]
+enum SyncResult {
+    /// All relish files are synced (their paths returned).
+    /// Metadata file is synced too (path not returned).
+    Success { synced: Vec<PathBuf> },
+    /// All relish files are synced (their paths returned).
+    /// Metadata file is not synced (path not returned).
+    MetadataSyncError { synced: Vec<PathBuf> },
+    /// Some relish files are not synced, some are (paths returned).
+    /// Metadata file is not synced (path not returned).
+    LayerSyncError {
+        synced: Vec<PathBuf>,
+        not_synced: Vec<PathBuf>,
+    },
+}
+
+/// Synchronizes given layers and metadata contents of a certain image.
+/// Relishes are always synced before metadata files are, the latter gets synced only if
+/// the rest of the files are successfully processed.
+#[allow(clippy::too_many_arguments)]
+async fn synchronize_layers<'a, P, S: 'static + RelishStorage<RelishStoragePath = P>>(
+    config: &'static PageServerConf,
+    max_concurrent_sync: usize,
+    relish_storage: &'a S,
+    layers: impl Iterator<Item = PathBuf>,
+    sync_operation: SyncOperation,
+    new_metadata: &'a TimelineMetadata,
+    tenant_id: ZTenantId,
+    timeline_id: ZTimelineId,
+) -> SyncResult {
+    let concurrent_operations_limit = Arc::new(Semaphore::new(max_concurrent_sync));
```

Comment:
Should this be per tenant or pageserver wide?

Reply:
That is indeed was planned to be pageserver wide, nice catch.